specreduce/extract.py

```diff
@@ -123,6 +123,34 @@ def _to_spectrum1d_pixels(fluxes):
                       flux=fluxes)
 
 
+def _align_along_trace(img, array_trace, disp_axis=1, crossdisp_axis=0):
+    """
+    Given an arbitrary trace ``array_trace`` (an np.ndarray), roll
+    all columns of ``nddata`` to shift the NDData's pixels nearest
+    to the trace to the center of the spatial dimension of the
+    NDData.
+    """
+    # TODO: this workflow does not support extraction for >2D spectra
+    if not (disp_axis == 1 and crossdisp_axis == 0):
+        # take the transpose to ensure the rows are the cross-disp axis:
+        img = img.T
+
+    n_rows = img.shape[0]
+
+    # indices of all columns, in their original order
+    rows = np.broadcast_to(np.arange(img.shape[0])[:, None], img.shape)
+    cols = np.broadcast_to(np.arange(img.shape[1]), img.shape)
+
+    # we want to "roll" each column so that the trace sits in
+    # the central row of the final image
+    shifts = array_trace.astype(int) - n_rows // 2
+
+    # we wrap the indices so we don't index out of bounds
+    shifted_rows = np.mod(rows + shifts[None, :], img.shape[0])
+
+    return img[shifted_rows, cols]
```

Comment:
is this cheaper or different than `np.roll`/`rollaxis`?

Reply:
It's different because [np.roll](https://numpy.org/doc/stable/reference/generated/numpy.roll.html) is written to support rolling an array over one axis by a constant amount everywhere on that axis (their "shift" parameter takes one integer shift per rolled axis).

Reply:
According to the docs, `shift` can be a tuple (one integer shift per row), is that not what you need?

Reply:
See my comment above – shift takes one integer shift per *axis*.

Reply:
Ah, I misread the docs as `shift` being the same length as the axis corresponding to `axis`, not the same length as the list of axes passed to `axis`.  I guess you'd have to iterate manually then and rebuild the whole image, which probably would be more expensive than your manual implementation.  Thanks for the explanation.